aten/src/ATen/native/Gesv.cpp

```diff
@@ -72,6 +78,107 @@ static void applyGesv(Tensor& b, Tensor& A, std::vector<int64_t> infos) {
   }
 }
 
+std::tuple<Tensor&,Tensor&> _gesv_single_out_cpu(
+    Tensor& sol, Tensor& lu,
+    const Tensor& self, const Tensor& A) {
+#ifndef USE_LAPACK
+  AT_ERROR("gesv: LAPACK library not found in compilation");
+#endif
+  int64_t bx = self.size(0);
+  int64_t by = (self.dim() == 1) ? 1 : self.size(1);
+  int64_t ax = A.size(0);
+  int64_t ay = A.size(1);
+  int info = 0;
+  Tensor temp_sol;
+  Tensor temp_lu;
+
+  /* Init to column major format. `sol` and `lu` need to be contiguous
+   * since we pass sol.data() and lu.data() to Lapack */
+  bool tc_sol = isTransposeContiguous(sol);
+  auto self_t = self.view({bx, by}).t();
+  if (!tc_sol &&
+      !sol.is_contiguous() &&
+      sol.dim() == 2 &&
+      sol.size(0) == bx &&
+      sol.size(1) == by) {
+    // eg. torch.gesv(.., out=(n[::2], ...))
+    temp_sol = self_t.clone();
+  }
+
+  if (!temp_sol.defined()) {
+    if (tc_sol) {
+      sol.t().resize_({by, bx});
+    } else {
+      sol.resize_({by, bx});
+    }
+    if (&self == &sol) {
+      // note: not a comprehensive check. Will fail for views
+      if (!tc_sol) {
+        sol.copy_(self_t.clone());
+      }
+    } else {
+      if (tc_sol) {
+        sol.t().copy_(self_t);
+      } else {
+        sol.copy_(self_t);
+      }
+    }
+  }
```

Comment:
I'm very confused. At this point, should `sol` be transposed or not? They kind of make sense to me but I am still not sure if they are correct, because so many things are controlled by `tc_sol` being `true` or `false` that it is difficult to keep track.

I think it would be better to make the code clearer. It would be better for future maintenance too. Can you write these if statements so that they are
```
if (tc_sol) {
  // everything up till here...
} else {
  // everything up till here...
}
```
or write out your strategy clearly in a comment block. Thanks.

Reply:
Additionally, I wonder if this kind of structure will make it clearer.

```c++
// gesv takes in two tensors and returns them in-place. so unless the output
// args are the same as input args, we will need to make a copy.

// these tensors will be the input and output args to the lapack call.
// they may be sol and lu, or may be a buffer, indicated by flags sol_buffer
// and lu_buffer.
Tensor sol_lapack_arg;
Tensor lu_lapack_arg;

bool sol_correct_shape = sol.size(0) == bx && sol.size(1) == by;
bool sol_buffer = false;

// now we detect if we need to use a buffer and/or resize the out arguments.
//
// our strategy is the same for each out arg. bellow we use (sol, self) pair
// for demonstration purpose. 
// 1. if sol has correct shape and is fortran contig
//    copy self into sol (if self is not sol)
//    call lapack gesv using sol
// 2. if sol has correct shape but isn't fortran contig
//    make a buffer, copy self into the buffer s.t. it is fortran contig
//    call lapack gesv using the buffer
//    copy the buffer content into sol
// 3. if sol has incorrect shape, then we have to resize
//    resize sol inplace, and make it fortran contig and holding content of self
//    call lapack gesv using sol

if (sol_correct_shape && tc_sol) {
    if (&sol != &self) {
        sol.copy_(self);
    }
    sol_lapack_arg = sol;
} else if (sol_correct_shape) {  // shape right but not fortran contiguous
    sol_lapack_arg = self.t().clone().t_(); // make it fortran contiguous
    sol_buffer = true;
} else {  // wrong shape, resize_
    // since shape is wrong, resize_ call will make it (fortran) contiguous
    sol_lapack_arg = sol.resize_({by, bx}).t_().copy_(self);  
}

// same for lu

AT_DISPATCH_FLOATING_TYPES(self.type(), "gesv", [&]{
    auto ipiv = at::empty({bx}, sol.options().dtype(kInt));
    lapackGesv<scalar_t>(bx, by, lu_lapack_arg.data<scalar_t>, bx, 
                         ipiv.data<int>(), sol_lapack_arg.data<scalar_t>, 
                         bx, &info);
    });
checkErrors({info});

if (sol_buffer) {
    sol.copy_(sol_lapack_arg);
}

// same for lu

return std::make_tuple(sol, lu);
```